heartbeat/skima/core.go

```diff
@@ -0,0 +1,168 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package skima
+
+import (
+	"strings"
+
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+
+	"fmt"
+
+	"sort"
+
+	"github.com/elastic/beats/libbeat/common"
+)
+
+// Is creates a named IsDef with the given checker.
+func Is(name string, checker Checker) IsDef {
+	return IsDef{name: name, checker: checker}
+}
+
+// Optional wraps an IsDef to mark the field's presence as optional.
+func Optional(id IsDef) IsDef {
+	id.name = "optional " + id.name
+	id.optional = true
+	return id
+}
+
+// Map is the type used to define schema definitions for Schema.
+type Map map[string]interface{}
+
+// SchemaValidator is the result of Schema and is run against the map you'd like to test.
+type SchemaValidator func(common.MapStr) SchemaValidationResults
+
+// Compose combines multiple SchemaValidators into a single one.
+func Compose(validators ...SchemaValidator) SchemaValidator {
+	return func(actual common.MapStr) SchemaValidationResults {
+		results := make([]SchemaValidationResults, len(validators))
+		for idx, validator := range validators {
+			results[idx] = validator(actual)
+		}
+
+		combined := SchemaValidationResults{}
+		for _, r := range results {
+			r.eachResult(func(path string, vr ValueResult) {
+				combined.recordResult(path, vr)
+			})
+		}
+		return combined
+	}
+}
+
+// Strict is used when you want any unspecified keys that are encountered to be considered errors.
+func Strict(laxValidator SchemaValidator) SchemaValidator {
+	return func(actual common.MapStr) SchemaValidationResults {
+		validatedResults := laxValidator(actual)
```

Comment:
everything is `validatedX` in here. Try to keep short names. E.g. code is understandable as well, by using `results, paths`.